File: python-algorithm/algo_05_dynamic_plan/dp_4_special_state_transition/L279-m.py
#--coding:utf-8--
'''
@Time   : 2020/10/12
@Author : No Name
'''

# 动态规划：可以看作是完全背包问题
class Solution:
    def numSquares(self, n: int) -> int:
        dp = [i for i in range(n+1)]
        dp[0] = 0
        dp[1] = 1
        for i in range(2, n+1):
            for j in range(1, int(sqrt(i))+1):
                dp[i] = min(dp[i], dp[i-j*j]+1)
        return dp[n]

# 宽搜
class Solution:
    def numSquares(self, n: int) -> int:
        visited = [0] * (n + 1)
        que = []
        que.append(0)
        level = 0
        while que != []:
            level += 1
            for i in range(len(que)):
                cur = que.pop(0)
                if cur == n:
                    return level - 1
                for j in range(1, int(sqrt(n)) + 1):
                    tmp = cur + j*j
                    if tmp <= n and not visited[tmp]:
                        que.append(tmp)
                        visited[tmp] = 1

File: python-algorithm/algo_05_dynamic_plan/dp_4_special_state_transition/L279-m.cpp


//使用的DP,另外一种方法是图遍历

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int numSquares(int n) {
        vector<int> dp(n+1, 0);
        dp[0] = 0;
        for(int i=1; i<=n; ++i){
            int s = 1;
            int min_num = i; //i个1  最开始这个代码放错位置了，必须放在循环里面
            while(i-s*s >= 0){
                min_num = min(min_num, dp[i-s*s] + 1); //注意啊 是i-s*s 不是n-s*s ，当前遍历求得是到整数i的平方数路径最短
                s++;
            }
            dp[i] = min_num;
        }
        return dp[n];
    }
};